crates/neon/src/instance/mod.rs

```diff
@@ -0,0 +1,124 @@
+use std::any::Any;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use once_cell::sync::OnceCell;
+
+use crate::context::Context;
+use crate::lifecycle::GlobalCell;
+use crate::result::NeonResult;
+
+static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+fn next_id() -> usize {
+    COUNTER.fetch_add(1, Ordering::SeqCst)
+}
+
+/// A cell that can be used to allocate data that is global to an instance
+/// of a Neon addon.
+#[derive(Default)]
+pub struct Global<T> {
+    _type: PhantomData<T>,
+    id: OnceCell<usize>,
+}
+
+impl<T> Global<T> {
+    /// Creates a new global value. This method is `const`, so it can be assigned to
+    /// static variables.
+    pub const fn new() -> Self {
+        Self {
+            _type: PhantomData,
+            id: OnceCell::new(),
+        }
```

Comment:
Oops, this failed to type check because `T` might not implement `Default` (annoying!). I reverted it.

Reply:
Oh, yea. Also, `Default::default` isn't a `const fn`. It makes more sense to go the other way and implement `Default` with `Global::new`.

```rust
pub struct Global<T> {
    _type: PhantomData<T>,
    id: OnceCell<usize>,
}

impl<T> Default for Global<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T> Global<T> {
    pub const fn new() -> Self {
        Self {
            _type: PhantomData,
            id: OnceCell::new(),
        }
    }
}
```

The derived `Default` will require `T: Default`, but if you manually implement, that bound isn't necessary.

Reply:
Oh, you're right. This would require removing `'cx` and replacing it with a `for<'cx>` *everywhere*. While that works, it's much simpler to have a `PhantomData`.